File: solutions/2026-02-28/python/2026-02-28-concatenation-of-consecutive-binary-numbers.py
```python
# Problem: Concatenation of Consecutive Binary Numbers
# Problem Summary: Concatenate binary representations of numbers from 1 to n and return the decimal value modulo 10^9 + 7.
# Link: https://leetcode.com/problems/concatenation-of-consecutive-binary-numbers/
#
# Approach:
# We can simulate the concatenation process. We maintain a running `result` which stores the decimal value of the concatenated binary string so far.
# For each number `i` from 1 to `n`, we need to determine how many bits its binary representation has.
# If `i` is a power of 2 (e.g., 1, 2, 4, 8), its binary representation has one more bit than the previous number's binary representation.
# We can check if `i` is a power of 2 by using the property `(i & (i - 1)) == 0`.
#
# Let `bits` be the number of bits in the binary representation of the current number `i`.
# When we process `i`, we essentially shift the current `result` to the left by `bits` positions (equivalent to multiplying by 2^bits) and then add `i`.
# So, the update rule becomes: `result = (result << bits) + i`.
# Since the result can be very large, we apply the modulo operation `10^9 + 7` at each step to prevent overflow.
#
# We can determine the number of bits for `i` efficiently. For example, if `i` is 3 (binary "11"), it has 2 bits. If `i` is 4 (binary "100"), it has 3 bits.
# Notice that the number of bits increases only when `i` becomes a power of 2.
# We can keep track of `bits` and increment it whenever `i` is a power of 2. A simpler way to find the number of bits for `i` is `i.bit_length()`.
#
# However, a more efficient way to determine `bits` for `i` is to observe that `bits` increases by 1 every time `i` crosses a power of 2.
# For example:
# i=1: "1" (1 bit)
# i=2: "10" (2 bits)
# i=3: "1011" (2 bits for 3)
# i=4: "1011100" (3 bits)
#
# Let's refine the `bits` calculation. We can maintain the `bits` count. When `i` becomes a power of 2 (e.g., 2, 4, 8), we increment `bits`.
# A number `i` is a power of 2 if `i` is `1 << k` for some integer `k`.
# Alternatively, we can use `i.bit_length()` directly for each `i`. This is cleaner.
#
# Let `MOD = 10**9 + 7`.
# Initialize `result = 0`.
# For `i` from 1 to `n`:
#   `num_bits = i.bit_length()`  # Get the number of bits for i
#   `result = (result << num_bits) % MOD`  # Shift the current result left by num_bits
#   `result = (result + i) % MOD`         # Add the current number i
# Return `result`.
#
# Time Complexity: O(n * log n) where log n is the average bit length of numbers up to n.
# Actually, `i.bit_length()` is O(log i). So the total time complexity is the sum of `log i` for i from 1 to n, which is roughly O(n log n).
# A tighter analysis might consider that `bit_length()` is proportional to the number of bits. The total number of bits concatenated is `sum(bit_length(i) for i in 1..n)`.
# The number of bits for `i` is approximately `log2(i)`. The sum `sum(log2(i))` from 1 to n is approximately `n log2(n)`.
# However, if we look at how `bit_length()` increases: it stays constant for ranges of numbers.
# For numbers from 1 to 1: 1 bit.
# For numbers from 2 to 3: 2 bits.
# For numbers from 4 to 7: 3 bits.
# For numbers from 2^k to 2^(k+1)-1: k+1 bits.
# The number of elements in each range is roughly `2^k`.
# The total number of operations is proportional to the total number of bits, which is approximately `n * log2(n)`.
#
# More precisely, if `L` is the number of bits for `n` (i.e., `n.bit_length()`), then the sum of bits is:
# `sum(k * 2^(k-1) for k=1 to L-1) + L * (n - 2^(L-1) + 1)`
# This sum is dominated by the last term, which is `L * n`. So, the time complexity is O(n * log n).
#
# Space Complexity: O(1) as we only use a few variables to store the result and loop counter.

class Solution:
    def concatenatedBinary(self, n: int) -> int:
        # Define the modulo constant
        MOD = 10**9 + 7

        # Initialize the decimal result to 0
        result = 0

        # Iterate through numbers from 1 to n
        for i in range(1, n + 1):
            # Determine the number of bits in the binary representation of i.
            # For example, 1 is "1" (1 bit), 2 is "10" (2 bits), 3 is "11" (2 bits), 4 is "100" (3 bits).
            num_bits = i.bit_length()

            # Shift the current result to the left by `num_bits` positions.
            # This is equivalent to multiplying by 2^num_bits.
            # We apply modulo at each step to prevent integer overflow.
            result = (result << num_bits) % MOD

            # Add the current number `i` to the shifted result.
            # Again, apply modulo to keep the result within bounds.
            result = (result + i) % MOD

        # Return the final concatenated binary value as a decimal, modulo 10^9 + 7
        return result

```

File: solutions/2026-02-28/cpp/2026-02-28-concatenation-of-consecutive-binary-numbers.cpp
```cpp
// Problem: Concatenation of Consecutive Binary Numbers
// Link: https://leetcode.com/problems/concatenation-of-consecutive-binary-numbers
//
// Approach:
// We can iterate from 1 to n. For each number `i`, we need to find its binary
// representation and append it to our growing binary string.
// However, constructing a large string and then converting it to a decimal
// will be inefficient and may lead to overflow.
//
// Instead, we can maintain a running decimal value. For each number `i`,
// we determine the number of bits required for its binary representation.
// Let's say `i` requires `k` bits. Then, to append the binary representation
// of `i` to the current decimal value `res`, we shift `res` left by `k`
// bits (effectively multiplying by 2^k) and then add `i`.
//
// To find `k` (the number of bits for `i`), we can use `log2(i)` or
// find the most significant bit. A simple way is to keep shifting `i`
// right until it becomes 0, counting the shifts. Or, more efficiently,
// we can observe that the number of bits for `i` increases only when `i`
// is a power of 2. We can track the number of bits needed and update it.
// For example, numbers 1-1 require 1 bit. Numbers 2-3 require 2 bits.
// Numbers 4-7 require 3 bits, and so on.
//
// We need to perform all calculations modulo 10^9 + 7 to prevent overflow.
//
// Time Complexity: O(n * log n) because for each number from 1 to n, we perform
// bitwise operations which take logarithmic time with respect to the number itself.
// More precisely, it's O(n * B) where B is the maximum number of bits for n,
// which is log(n).
//
// Space Complexity: O(1) as we are only using a few variables to store the result
// and intermediate calculations.
//

class Solution {
public:
    int concatenatedBinary(int n) {
        long long res = 0; // Stores the concatenated decimal value
        int MOD = 1e9 + 7; // The modulo value
        int num_bits = 0;  // Tracks the number of bits needed for the current number

        for (int i = 1; i <= n; ++i) {
            // If 'i' is a power of 2, it means we are starting a new block of
            // binary numbers that will require one more bit than the previous block.
            // For example, when i=2 (binary "10"), we need 2 bits. When i=4 (binary "100"),
            // we need 3 bits.
            // We can detect a power of 2 by checking if (i & (i - 1)) == 0.
            if ((i & (i - 1)) == 0) {
                num_bits++; // Increment the number of bits required
            }

            // Append the binary representation of 'i' to the result.
            // To do this, we first shift the current result 'res' to the left
            // by 'num_bits'. This effectively makes space for the new binary
            // representation of 'i'.
            // Then, we add 'i' to the shifted result.
            // All operations are performed modulo MOD.
            res = (res << num_bits) % MOD; // Shift left by num_bits and take modulo
            res = (res + i) % MOD;         // Add 'i' and take modulo
        }

        return (int)res; // Cast the final result back to int
    }
};
```